### PR TITLE
OpStack finality v2

### PR DESCRIPTION
**Pull request by michalsidzej**
Resolves L2B-4275

## New .env variables  

```
ETHEREUM_BEACON_API_URL=
ETHEREUM_BEACON_API_CALLS_PER_MINUTE=
```
Beacon api url can be the same as rpc url from quicknode.

## Context

This PR introduces decoding of blob data opStack chains to calculate finality. It's still not perfect logic, as we ignore frames that are posted across different batches, but taking it under the consideration requires a different approach to finality.

There's a lot of code that goes over relatively large pieces of data (blobs). I have to say this is my first time working with similar variables, so please don't hesitate to comment on possible improvements. 

## Next steps

Before merging this PR I will disable automatic promote to production on heroku to test if this process is not too memory-intensive

### CODE REVIEW

**Review by mateuszradomski — CHANGES_REQUESTED** [review]

File: packages/backend/src/modules/finality/analyzers/opStack/decodeSpanBatch.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import { assert } from '@l2beat/backend-tools'
+import { UnixTime } from '@l2beat/shared-pure'
+import proto from 'protobufjs'
+
+const SPAN_BATCH_VERSION = 1
+
+export interface SpanBatchDecoderOpts {
+  l2BlockTimeSeconds: number
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/decodeSpanBatch.ts:8]: I strongly disagree. https://www.investopedia.com/terms/b/block-time-cryptocurrency.asp#:~:text=Block%20time%20measures%20the%20time,it%20was%20introduced%20in%202009. https://docs.optimism.io/chain/differences#:~:text=Block%20time%20in%20seconds

[the remark this answers is not in the document]
Comment (mateuszradomski) [reply packages/backend/src/modules/finality/analyzers/opStack/decodeSpanBatch.ts:8]: Well, I'm not going to be arguing against a standard :p Consider this comment not relevant anymore

File: packages/backend/src/modules/finality/analyzers/opStack/ChannelBank.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  frameNumber: number
+}
+
+type channelId = string
+
+export class ChannelBank {
+  // TODO: channels should be saved to the db so they're persistent
+  // In the opStack spec, channels are recorded in FIFO order in a structure called the channel queue.
+  // A channel is added to the channel queue the first time a frame belonging to the channel is seen.
+  // Because we do not look at all the transactions from opStack chains, we ignore this FIFO assumption (could be improved)
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/ChannelBank.ts:19]: [yes sir](https://linear.app/l2beat/issue/L2B-4596/decode-the-proper-channel-data-for-opstack)

File: packages/backend/src/modules/finality/analyzers/opStack/blobToData.ts
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/blobToData.ts:1]: [19d1e94](https://github.com/l2beat/l2beat/pull/3193/commits/19d1e944f7248b80b33aad21ddcf88034dc16634)

File: packages/backend/src/modules/finality/analyzers/opStack/blobToData.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  )
+
+  // round 0 is special cased to copy only the remaining 27 bytes of the first field element into
+  // the output due to version/length encoding already occupying its first 5 bytes.
+  const output = new Uint8Array(MaxBlobDataSize)
+  output.set(blob.subarray(5, 32), 0)
+
+  // now process remaining 3 field elements to complete round 0
+  let opos = 28 // current position into output buffer
+  let ipos = 32 // current position into the input blob
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/blobToData.ts]: [19d1e94](https://github.com/l2beat/l2beat/pull/3193/commits/19d1e944f7248b80b33aad21ddcf88034dc16634)

File: packages/backend/src/modules/finality/analyzers/opStack/getFrames.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+const FIXED_OVERHEAD_BYTES = 23
+
+// https://specs.optimism.io/protocol/derivation.html#batcher-transaction-format
+export function getFrames(rollupData: Uint8Array) {
+  assert(
+    rollupData[VERSION_OFFSET] === ENCODING_VERSION,
+    `Invalid version received, expected ${ENCODING_VERSION} received ${rollupData[VERSION_OFFSET]}`,
+  )
+
+  const frames = rollupData.slice(1)
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/getFrames.ts:15]: [19d1e94](https://github.com/l2beat/l2beat/pull/3193/commits/19d1e944f7248b80b33aad21ddcf88034dc16634)

File: packages/backend/src/modules/finality/analyzers/opStack/getFrames.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+function byteArrToNum(arr: Uint8Array): number {
+  assert(arr.length > 0)
+
+  if (arr.length === 1) {
+    return arr[0]
+  }
+
+  return 256 * byteArrToNum(arr.slice(0, arr.length - 1)) + arr[arr.length - 1]
+}
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/getFrames.ts]: [19d1e94](https://github.com/l2beat/l2beat/pull/3193/commits/19d1e944f7248b80b33aad21ddcf88034dc16634)

File: packages/backend/src/modules/finality/analyzers/opStack/ChannelBank.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/ChannelBank.ts]: [19d1e94](https://github.com/l2beat/l2beat/pull/3193/commits/19d1e944f7248b80b33aad21ddcf88034dc16634)

File: packages/backend/src/modules/finality/analyzers/opStack/getFrames.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  // uint16
+  const frameNumber = byteArrToNum(frames.slice(16, 18))
+  // uint32
+  const frameDataLength = byteArrToNum(frames.slice(18, 22))
+  // bytes
+  const frameData = frames.slice(22, 22 + frameDataLength)
+  // bool
+  const isLast = !!byteArrToNum(
+    frames.slice(22 + frameDataLength, 22 + frameDataLength + 1),
+  )
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/getFrames.ts]: [1c77105](https://github.com/l2beat/l2beat/pull/3193/commits/1c771052d1007bdb28bb60d790fd4f2adeb4cdc1)

File: packages/backend/src/modules/finality/analyzers/opStack/utils.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  for (const chunk of chunks) {
+    concatenatedChunks.set(chunk, offset)
+    offset += chunk.length
+  }
+  return concatenatedChunks
+}
+
+export function byteArrFromHexStr(hexString: string) {
+  const str = hexString.startsWith('0x') ? hexString.slice(2) : hexString
+  const match = str.match(/.{1,2}/g)
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/utils.ts]: [f12a86c](https://github.com/l2beat/l2beat/pull/3193/commits/f12a86c1cd9905b62137ee07b7390b3ef12e74e6)

File: packages/backend/src/modules/finality/analyzers/opStack/decodeSpanBatch.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
+import { assert } from '@l2beat/backend-tools'
+import { UnixTime } from '@l2beat/shared-pure'
+import proto from 'protobufjs'
```
[the remark this answers is not in the document]
Comment (michalsidzej) [reply packages/backend/src/modules/finality/analyzers/opStack/decodeSpanBatch.ts]: [f12a86c](https://github.com/l2beat/l2beat/pull/3193/commits/f12a86c1cd9905b62137ee07b7390b3ef12e74e6)

**Review by mateuszradomski — APPROVED** [review]
lgtm

**Pull request merged**